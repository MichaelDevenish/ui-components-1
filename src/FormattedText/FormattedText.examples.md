# Formatted Text

FormattedText is used to display preformatted text from a database.
It preserves whitespace and renders URLs as links.

```jsx
  <FormattedText
    text={`Hey this is the formatted text component

Here is a link: www.intellihr.com.au`}
  />
```
