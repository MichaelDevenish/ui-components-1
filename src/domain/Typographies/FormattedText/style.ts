import React from 'react'
import styled, { StyledComponentClass } from 'styled-components'

export const StyledFormattedText = styled.div`
  white-space: pre-wrap;

  .react-markdown-block {
    > :last-child {
      margin-bottom: 0;
    }
  }
`
