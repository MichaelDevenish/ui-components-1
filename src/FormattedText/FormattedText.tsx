import React from 'react'
import ReactMarkdown from 'react-markdown'
import { Emoji } from 'emoji-mart'
import reactStringReplace from 'react-string-replace'
import { StyledFormattedText } from './style'
import { TextLink } from '../Link'

export interface FormattedTextProps {
  /** Preformatted markdown text to display */
  text: string
}

export class FormattedText extends React.PureComponent<FormattedTextProps> {
  linkRenderer = (
    props: { children: any, href: string }
  ): JSX.Element => {
    return (
      <TextLink
        href={props.href}
        target='_blank'
        useReactRouter={false}
      >
        {props.children}
      </TextLink>
    )
  }

  textRenderer = (
    text: string
  ): JSX.Element => {
    return reactStringReplace(text, /:([^:]+):/g, (match: string, i: number) => {
      return <Emoji
        key={i}
        emoji={match}
        set='twitter'
        size={16}
      />
    })
  }

  public render (): JSX.Element {
    const {
      text
    } = this.props

    return (
      <StyledFormattedText>
        <ReactMarkdown
          source={text}
          renderers={{
            link: this.linkRenderer,
            text: this.textRenderer
          }}
        />
      </StyledFormattedText>
    )
  }
}
