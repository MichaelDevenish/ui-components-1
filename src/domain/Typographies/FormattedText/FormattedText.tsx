import { Emoji } from 'emoji-mart'
import React from 'react'
import ReactMarkdown from 'react-markdown'
import reactStringReplace from 'react-string-replace'

import { TextLink } from '../../Links'
import { StyledFormattedText } from './style'

export interface IFormattedTextProps {
  /** Preformatted markdown text to display */
  text: string
  /** If true, renders emojis */
  renderEmojis?: boolean
}

export class FormattedText extends React.PureComponent<IFormattedTextProps> {
  public static defaultProps: Partial<IFormattedTextProps> = {
    renderEmojis: false
  }

  public linkRenderer = (props: { children: any, href: string }): JSX.Element => {
    return (
      <TextLink
        anchorComponent='a'
        href={props.href}
        target='_blank'
      >
        {props.children}
      </TextLink>
    )
  }

  public textRenderer = (text: string): JSX.Element => {
    return reactStringReplace(text, /(?:^|\s):([^:\s]+):(?=\s|$)/g, (match: string, i: number) => {
      return (
        <Emoji
          key={i}
          emoji={match}
          set='twitter'
          size={16}
        />
      )
    })
  }

  public render (): JSX.Element {
    const {
      text,
      renderEmojis
    } = this.props

    const renderers: any = {
      link: this.linkRenderer
    }

    if (renderEmojis) {
      renderers.text = this.textRenderer
    }

    return (
      <StyledFormattedText>
        <ReactMarkdown
          source={text}
          renderers={renderers}
          className='react-markdown-block'
        />
      </StyledFormattedText>
    )
  }
}
