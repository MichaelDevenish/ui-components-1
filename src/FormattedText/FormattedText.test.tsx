import { mount } from 'enzyme'
import React from 'react'
import { FormattedText } from '../FormattedText'

describe('<FormattedText />', () => {
  it(`should render the formatted text`, () => {
    const wrapper = mount(
      <FormattedText
        text='Hello. I am a piece of paragraph text.

        This should be on a new line :)'
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it(`should render any urls as links`, () => {
    const wrapper = mount(
      <FormattedText
        text='www.heyguy.com

        also this one www.ozbargain.com.au'
      />
    )

    expect(wrapper).toMatchSnapshot()
  })
})
