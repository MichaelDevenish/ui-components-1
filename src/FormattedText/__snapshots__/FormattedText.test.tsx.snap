// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormattedText /> should render any urls as links 1`] = `
.c0 {
  white-space: pre-wrap;
}

<FormattedText
  text="www.heyguy.com

        also this one www.ozbargain.com.au"
>
  <styled.div>
    <div
      className="c0"
    >
      <Linkify
        className="Linkify"
        component="a"
        properties={
          Object {
            "target": "_blank",
          }
        }
      >
        <span
          className="Linkify"
        >
          <a
            href="http://www.heyguy.com"
            key="parse0match0"
            target="_blank"
          >
            www.heyguy.com
          </a>
          

        also this one 
          <a
            href="http://www.ozbargain.com.au"
            key="parse0match1"
            target="_blank"
          >
            www.ozbargain.com.au
          </a>
        </span>
      </Linkify>
    </div>
  </styled.div>
</FormattedText>
`;

exports[`<FormattedText /> should render the formatted text 1`] = `
.c0 {
  white-space: pre-wrap;
}

<FormattedText
  text="Hello. I am a piece of paragraph text.

        This should be on a new line :)"
>
  <styled.div>
    <div
      className="c0"
    >
      <Linkify
        className="Linkify"
        component="a"
        properties={
          Object {
            "target": "_blank",
          }
        }
      >
        <span
          className="Linkify"
        >
          Hello. I am a piece of paragraph text.

        This should be on a new line :)
        </span>
      </Linkify>
    </div>
  </styled.div>
</FormattedText>
`;
